### PR TITLE
Add safer updates to collection view changes

### PR DESCRIPTION
**Pull request by nuno-vieira**
## Description of the pull request
Fixes potential crashes when there are collection view changes. Checks if there are any conflicting changes, if there is, reloads the collection data, instead of performing batch updates.

### CODE REVIEW

File: Sources/StreamChatUI/ChatChannelList/ChatChannelListVC.swift [hunk 1]
```diff
@@ -233,31 +233,65 @@ extension _ChatChannelListVC: _ChatChannelListControllerDelegate {
...
+        // If doing an union of all IndexPath changes and the count is less than before
+        // It means there are duplicates, and so we have conflicting updates
+        let hasConflictingUpdates = mergedItems.count != numberOfUpdates
+
+        if hasConflictingUpdates {
+            collectionView.reloadData()
+
+            channelsCount = controller.channels.count
+            return
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatChannelList/ChatChannelListVC.swift:278]: Can you move all this to a separate method/extension? Also, if the updates are conflicting updates, can we do something like:
```
log.error("⚠️ Inconsistent updates from ChatChannelListController. Please copy the following log and open an issue at https://github.com/GetStream/stream-chat-swift/issues")
+ pretty print the changes
```

**Review by VojtaStavik — APPROVED** [review]
This is so far the most sophisticated approach to this issue. Let's use it to gather as much info about the issue as possible.

**Review by olejnjak — APPROVED** [review]
🤯

**Comment by nuno-vieira** [comment]
> 🤯

@olejnjak Can you please re-review again? 😄 I think now is more readable and easier to understand and also more optimized. But this is a temporary solution anyways :)

File: Sources/StreamChatUI/ChatChannelList/ChatChannelListVC.swift [hunk 2]
```diff
@@ -233,31 +233,70 @@ extension _ChatChannelListVC: _ChatChannelListControllerDelegate {
...
+                removeIndexes.insert(index)
+            case let .update(_, index):
+                verifyConflict(index)
+                updateIndexes.insert(index)
+            }
+        }
+
+        if hasConflicts {
+            collectionView.reloadData()
+            channelsCount = controller.channels.count
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatChannelList/ChatChannelListVC.swift:274]: this has to be the other way around - first update the count, then reload the view

**Review by VojtaStavik — APPROVED** [review]

**Comment by olejnjak** [comment]
> > 🤯
> 
> @olejnjak Can you please re-review again? 😄 I think now is more readable and easier to understand and also more optimized. But this is a temporary solution anyways :)

That was more like it is insane that something like this is necessary 🤦 so more because of the cause, not the solution 🙂

**Review by olejnjak — APPROVED** [review]

**Comment by nuno-vieira** [comment]
> > > 🤯
> > 
> > 
> > @olejnjak Can you please re-review again? 😄 I think now is more readable and easier to understand and also more optimized. But this is a temporary solution anyways :)
> 
> That was more like it is insane that something like this is necessary 🤦 so more because of the cause, not the solution 🙂

Yes it is said :/ Thank you 🙂

**Pull request merged**